### ISSUE TITLE
tf.keras.Model.fit() training works fine but custom training loop fails for identical model, optimizer, and loss function

### ISSUE THREAD

**Issue by Georji-Ghosh** [tensorflow/tensorflow#46782]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10, Anaconda
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.4.0
- Python version: 3.8.0
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

v2.4.0-rc4-71-g582c8d236cb 2.4.0

**Describe the current behavior**
Code
`from tensorflow.python.eager import backprop
import tensorflow as tf
import numpy as np
import math

print("TensorFlow version: {}".format(tf.__version__))
print("Eager execution: {}".format(tf.executing_eagerly()))

#Equation to generate data
#y = x * 0.1
def calulate(x):
    return sum([i*0.1 for i in x])

#Generate samples in batch
def samples(size):
    x = np.random.random((size, 4))
    y = np.zeros((size))
    for i in range(size):
        y[i] = calulate(x[i])
    return x,y

#Model
model = tf.keras.Sequential([
  tf.keras.layers.Dense(1, input_shape=(4,))
])

#loss function and optimizer
loss = tf.keras.losses.MeanSquaredError(reduction="auto")
optimizer = tf.keras.optimizers.Adam(learning_rate=0.01)

#Validation function
def validate():
    x = np.random.random((5, 4))
    y = [calulate(i) for i in x]
    e = model.predict(x)
    print("{} ==> {} , {}".format(x, y, e))

#Model.fit
def train_default(epoch) :
    x, y = samples(64000)
    model.compile(optimizer=optimizer, loss=loss)
    history = model.fit(x, y, epochs=epoch, validation_split = 0.2)

#Custom train
def train_custom(steps):
    x = []
    y = []
    for _ in range(2000):
        xe, ye = samples(32)
        x.append(xe)
        y.append(ye)

    for episod in range(steps):
        episod_loss = tf.keras.metrics.Mean()
        for i in range(2000):
            with tf.GradientTape() as tape:
                e = model(x[i], training=True)
                l = loss(y[i], e)
            grads = tape.gradient(l, model.trainable_weights)
            #optimizer.minimize(l, model.trainable_variables, tape=tape)
            optimizer.apply_gradients(zip(grads, model.trainable_weights))
            episod_loss.update_state(l)
        print("episod: {} loss: {}".format(episod, episod_loss.result()))
        print(model.trainable_weights)

#Tensorflow extension
loss_tracker = tf.keras.metrics.Mean(name="loss")
mae_metric = tf.keras.metrics.MeanAbsoluteError(name="mae")

class CustomModel(tf.keras.Model):
    def train_step(self, data):
        x, y = data

        with tf.GradientTape() as tape:
            y_pred = self(x, training=True)  # Forward pass
            l = loss(y, y_pred)

        #Compute gradients
        trainable_vars = self.trainable_variables
        gradients = tape.gradient(l, trainable_vars)

        #Update weights
        self.optimizer.apply_gradients(zip(gradients, trainable_vars))

        #Compute our own metrics
        loss_tracker.update_state(l)
        mae_metric.update_state(y, y_pred)
        return {"loss": loss_tracker.result(), "mae": mae_metric.result()}

    @property
    def metrics(self):
        return [loss_tracker, mae_metric]

def train_extended(steps):
    #Construct an instance of CustomModel
    inputs = tf.keras.Input(shape=(4,))
    outputs = tf.keras.layers.Dense(1)(inputs)
    model = CustomModel(inputs, outputs)

    #We don't passs a loss or metrics here.
    model.compile(optimizer=optimizer)

    x = []
    y = []
    for _ in range(2000):
        xe, ye = samples(32)
        x.append(xe)
        y.append(ye)

    for episod in range(steps):
        for i in range(2000):
            model.fit(x[i], y[i], epochs=1, verbose= 0)
        print("episod: {}".format(episod))
        print(model.trainable_variables)

##Works
#train_default(5)
#validate()

#Works
#train_extended(5)
#validate()

#Don't work
train_custom(5)
validate()`
**Describe the expected behavior**
train_custom function should also works like other 
**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
@Georji-Ghosh,
I did not face any errors while running the code with TF v2.4. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/fa5666b41e856603f5260dd6f343fef3/46782.ipynb). 

Could you please let us know the expected output, so that we can look into it. Thanks!

**Comment by Georji-Ghosh** [comment]
@amahendrakar Thanks for taking time to look into the issue.
The code runs properly and there is no issue in execution. But the issue is in the result generated.
1. When we train the model from scratch using either of "train_default()" or "train_extended()"  methods it works fine and all the weights goes ~0.1 as expected as the data generated using y=0.1x method and model has only one dense layer mapping input to output.
`episod: 4
[<tf.Variable 'dense_2/kernel:0' shape=(4, 1) dtype=float32, numpy=
array([[0.10025591],
       [0.09995212],
       [0.10002392],
       [0.10034888]], dtype=float32)>, <tf.Variable 'dense_2/bias:0' shape=(1,) dtype=float32, numpy=array([0.00022549], dtype=float32)>]`
2. When we train the model using custom training loop "train_custom()" it doesn't converges and the weights gets wrong values.
`episod: 4 loss: 0.0034574554301798344
[<tf.Variable 'dense_1/kernel:0' shape=(4, 1) dtype=float32, numpy=
array([[0.00126073],
       [0.00269012],
       [0.00523321],
       [0.0034958 ]], dtype=float32)>, <tf.Variable 'dense_1/bias:0' shape=(1,) dtype=float32, numpy=array([0.19942063], dtype=float32)>]`

My expectation is that both ways of training with same amount of training data should generate the same kind of weights in the model. 

actually to be more specific model.fit is not working same as

           with tf.GradientTape() as tape:
                e = model(x[i], training=True)
                l = loss(y[i], e)
            grads = tape.gradient(l, model.trainable_weights)
            #optimizer.minimize(l, model.trainable_variables, tape=tape)
            optimizer.apply_gradients(zip(grads, model.trainable_weights))
            episod_loss.update_state(l)

**Comment by Georji-Ghosh** [comment]
Can someone please help me understand why the custom training has a different set of weights after training then model.fit training?  or am I missing some points here?

**Comment by amahendrakar** [comment]
@ymodak,
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/abb333c21663dcda56e0b4b896907b99/46782.ipynb). Thanks!

**Comment by ymodak** [comment]
@Georji-Ghosh Please take a look at [SO thread](https://stackoverflow.com/questions/63550752/tf2-gradienttape-vs-model-fit-why-does-gradienttape-doesnt-work) discussion and see if it helps. Thanks!

**Comment by Georji-Ghosh** [comment]
Thanks for your input it resolve the issue for me I missed to convert the numpy array to tensor before sending it to the model and loss function.
